Block/BlockContextManager.php

```diff
@@ -214,4 +212,31 @@ protected function setDefaultExtraCacheKeys(BlockContextInterface $blockContext,
             $blockContext->setSetting('extra_cache_keys', $extraCacheKeys);
         }
     }
+
+    private function resolve(BlockInterface $block, $settings)
+    {
+        $optionsResolver = new OptionsResolver();
+
+        $this->configureSettings($optionsResolver, $block);
+
+        $service = $this->blockService->get($block);
+        $service->setDefaultSettings($optionsResolver, $block);
+
+        $reflector = new \ReflectionMethod($service, 'setDefaultSettings');
```

Comment:
How a `CompilerPass` will help to detect and throw deprecation errors?

I did the same method as Symfony. Please see: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Form/ResolvedFormType.php#L201-L207